### PR TITLE
Build asset for the Interoperability Module

### LINKED ISSUE #79: Build asset for the Interoperability Module
### Description

PR https://github.com/LiskHQ/lips/pull/208 updates [LIP-0063](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0063.md) and introduces a new method to build the asset for the Interoperability module.

### Acceptance Criteria

- New constants are added:
  - `EMPTY_BYTES`
  - `MODULE_NAME_INTEROPERABILITY`
  - `CHAIN_NAME_MAINCHAIN`
- The following methods are implemented:
   - `addInteropModuleEntry`
- The following methods are updated, wherever applicable:
  - `addLegacyModuleEntry`
  - `createLegacyReserveAccount`
  - `getLockedBalances`
  - `addAuthModuleEntry`
  - `addPoSModuleEntry`
  - `getStakes`
  - `createStakersArray`
- Tests are up-to-date

### PR DESCRIPTION
**Pull request by nagdahimanshu**
### What was the problem?

This PR resolves #79 

### How was it solved?
- [x] New constants are added:
  - [x] EMPTY_BYTES
  - [x] MODULE_NAME_INTEROPERABILITY
  - [x] CHAIN_NAME_MAINCHAIN
- [x] The following methods are implemented:
  - [x] addInteropModuleEntry
- [x] The following methods are updated, wherever applicable:
  - [x] addLegacyModuleEntry
  - [x] createLegacyReserveAccount
  - [x] getLockedBalances
  - [x] addAuthModuleEntry
  - [x] addPoSModuleEntry
  - [x] getStakes
  - [x] createStakersArray
- [x] Add/Update tests

### How was it tested?
Local

### CODE REVIEW

File: test/unit/assets/interoperability.spec.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+} from '../../../src/constants';
+import {
+	GenesisAssetEntry,
+	OwnChainDataSubstore,
+	RegisteredNamesSubstore,
+} from '../../../src/types';
+
+describe('Build assets/interoperability', () => {
+	const tokenID = '0400000000000000';
+	const chainID = tokenID.slice(0, 8);
```
Comment (priojeetpriyom) [inline test/unit/assets/interoperability.spec.ts:28]: same here (constant)

File: test/unit/assets/interoperability.spec.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+	it('should create interoperability module asset', async () => {
+		const response: GenesisAssetEntry = await addInteropModuleEntry(tokenID);
+		const ownChainDataSubstore = (response.data
+			.ownChainDataSubstore as unknown) as OwnChainDataSubstore;
+		const registeredNamesSubstore = (response.data
+			.registeredNamesSubstore as unknown) as RegisteredNamesSubstore;
```
Comment (priojeetpriyom) [inline test/unit/assets/interoperability.spec.ts:35]: Could it be directly typecasted without the `unknown`? If this is possible, please check for other places as well.

File: src/assets/pos.ts [hunk 3]
```diff
@@ -42,18 +40,24 @@ import {
 	Stake,
 	PoSStoreEntry,
 } from '../types';
-import { genesisPoSSchema } from '../schemas';
+
+const celing = (a: number, b: number) => {
```
Comment (sameersubudhi) [inline src/assets/pos.ts:44]:
```suggestion
const ceiling = (a: number, b: number) => {
```

File: src/assets/pos.ts [hunk 4]
```diff
@@ -158,12 +164,13 @@ export const createGenesisDataObj = async (
 	accounts: Account[],
```
Comment (sameersubudhi) [inline src/assets/pos.ts:164]: Please sort the array lexicographically (by `stakers.address`) before returning `stakers` on L160 within the `createStakersArray` method.

File: test/unit/assets/interoperability.spec.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+import { addInteropModuleEntry } from '../../../src/assets/interoperability';
+import {
+	CHAIN_NAME_MAINCHAIN,
```
Comment (sameersubudhi) [inline test/unit/assets/interoperability.spec.ts:16]:
```suggestion
	CHAIN_ID_LENGTH,
	CHAIN_NAME_MAINCHAIN,
```

File: test/unit/assets/interoperability.spec.ts (hunk 1 above)
Comment (sameersubudhi) [inline test/unit/assets/interoperability.spec.ts:28]:
```suggestion
	const chainID = tokenID.slice(0, CHAIN_ID_LENGTH * 2);
```

File: src/createAsset.ts [hunk 6]
```diff
@@ -51,25 +55,59 @@ export const getDataFromDBStream = async (stream: NodeJS.ReadableStream, schema:
...
-	const ids = await db.get(`${DB_KEY_TRANSACTIONS_BLOCK_ID}:${keyString(blockID)}`);
-	const idLength = 32;
-	for (let i = 0; i < ids.length; i += idLength) {
-		txIDs.push(ids.slice(i, i + idLength));
+export const getBlockHeadersByIDs = async (
+	db: KVStore,
+	arrayOfBlockIds: ReadonlyArray<Buffer>,
+): Promise<Buffer[]> => {
+	const blocks = [];
+	for (const id of arrayOfBlockIds) {
```
Comment (sameersubudhi) [inline src/createAsset.ts:63]:
```suggestion
	arrayOfBlockIDs: ReadonlyArray<Buffer>,
): Promise<Buffer[]> => {
	const blocks = [];
	for (const id of arrayOfBlockIDs) {
```

File: src/createAsset.ts [hunk 7]
```diff
@@ -51,25 +55,59 @@ export const getDataFromDBStream = async (stream: NodeJS.ReadableStream, schema:
...
+	return blockIDs;
+};
+
+export const getTransactions = async (blockID: Buffer, db: KVStore) => {
+	try {
+		const txIDs: Buffer[] = [];
+		const ids = await db.get(`${DB_KEY_TRANSACTIONS_BLOCK_ID}:${keyString(blockID)}`);
+		const idLength = 32;
+		for (let i = 0; i < ids.length; i += idLength) {
+			txIDs.push(ids.slice(i, i + idLength));
```
Comment (sameersubudhi) [inline src/createAsset.ts:93]: Where `TRANSACTION_ID_LENGTH = SHA_256_HASH_LENGTH` in `src/constant.ts`.
```suggestion
		const txIDLength = TRANSACTION_ID_LENGTH;
		for (let i = 0; i < ids.length; i += txIDLength) {
			txIDs.push(ids.slice(i, i + txIDLength));
```

File: src/createAsset.ts [hunk 8]
```diff
@@ -108,32 +148,23 @@ export class CreateAsset {
...
-				});
-
-				return {
-					blockHeader,
-					payload: decodedTransactions,
-				};
-			}),
-		);
+		const arrayOfBlockIds = await getBlocksIDsFromDBStream(blocksStream);
+		const blocksHeader = await getBlockHeadersByIDs(this._db, arrayOfBlockIds);
```
Comment (sameersubudhi) [inline src/createAsset.ts:152]: Please rename `arrayOfBlockIds` to `blockIDs` wherever applicable.
```suggestion
		const blockIDs: Buffer[] = await getBlocksIDsFromDBStream(blocksStream);
		const blocksHeader = await getBlockHeadersByIDs(this._db, blockIDs);
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: src/constants.ts [hunk 9]
```diff
@@ -67,3 +69,9 @@ export const DEFAULT_HOST = '127.0.0.1';
 export const DEFAULT_PORT_P2P = 7667;
 export const DEFAULT_PORT_RPC = 7887;
 export const KEEP_EVENTS_FOR_HEIGHTS = 300;
+export const HASH_LENGTH = 32;
+export const MAX_NUM_VALIDATORS = 199;
```
[the remark this answers is not in the document]
Comment (sameersubudhi) [reply src/constants.ts:73]: Please ignore this remark!

File: src/createAsset.ts [hunk 10]
```diff
@@ -51,25 +57,59 @@ export const getDataFromDBStream = async (stream: NodeJS.ReadableStream, schema:
...
+			const tx = await db.get(`${DB_KEY_TRANSACTIONS_ID}:${keyString(txID)}`);
+			const decodedTransactions = (await codec.decode(transactionSchema, tx)) as Transaction;
+
+			const id = hash(tx);
+			transactions.push({ ...decodedTransactions, id });
+		}
+
+		return transactions;
+	} catch (error) {
 		return [];
```
Comment (sameersubudhi) [inline src/createAsset.ts:111]: Can we please log the error (warn) and then return?

File: src/constants.ts [hunk 11]
```diff
@@ -69,9 +69,11 @@ export const DEFAULT_HOST = '127.0.0.1';
 export const DEFAULT_PORT_P2P = 7667;
```
Comment (sameersubudhi) [inline src/constants.ts:69]: Please double check the value for `Q96_ZERO` with the SDK/research team. Seems incorrect.

**Review by sameersubudhi — CHANGES_REQUESTED** [review]
Please consider the following. In addition, kindly remove the usage of `any` from all the places as discussed except `src/utils/genesis_block.ts`, which needs a TODO.

File: src/constants.ts [hunk 12]
```diff
@@ -17,28 +17,37 @@ export const MODULE_NAME_LEGACY = 'legacy';
...
+export const ADDRESS_LEGACY_RESERVE = hash(Buffer.from('legacyReserve', 'utf8')).slice(0, 20);
+export const INVALID_BLS_KEY = Buffer.alloc(48, 0).toString('hex');
+export const INVALID_ED25519_KEY = Buffer.alloc(32, 255).toString('hex');
+export const DUMMY_PROOF_OF_POSSESSION = Buffer.alloc(96, 0).toString('hex');
+
+export const CHAIN_ID_LENGTH = 4;
+export const CHAIN_NAME_MAINCHAIN = 'lisk_mainchain';
+
+export const Q96_ZERO = '';
+export const MAX_NUMBER_BYTES_Q96 = Buffer.alloc(0);
```
Comment (sameersubudhi) [inline src/constants.ts:50]:
```suggestion
export const Q96_ZERO = Buffer.alloc(0);
export const MAX_NUMBER_BYTES_Q96 = 24;
```

File: src/index.ts [hunk 13]
```diff
@@ -274,7 +274,7 @@ class LiskMigrator extends Command {
 				cli.action.start('Starting lisk-core v4');
 				try {
 					const network = networkConstant?.name as string;
-					await startLiskCore('PASS THE CONFIGURATION PATH', appVersion, client, { network });
+					await startLiskCore(configCoreV4, appVersion, client, { network });
```
Comment (sameersubudhi) [inline src/index.ts:277]: Please add a TODO here to verify the implementation properly.

**Review by sameersubudhi — APPROVED** [review]
LGTM 👍

File: src/assets/pos.ts [hunk 14]
```diff
@@ -41,19 +39,26 @@ import {
 	Staker,
 	Stake,
 	PoSStoreEntry,
+	StakerBuffer,
 } from '../types';
-import { genesisPoSSchema } from '../schemas';
+
+const ceiling = (a: number, b: number) => {
+	if (b === 0) throw new Error('Can not divide by 0.');
+	return Math.floor((a + b - 1) / b);
```
Comment (priojeetpriyom) [inline src/assets/pos.ts:47]: though not needed for current usage, but could we use `Math.ceil` instead? This won't work incase b <= 1. Ex: a=5.5, b=1. 
If we can not use `Math.ceil`, then maybe we can substract a smaller value like 1e-4?

File: src/createAsset.ts [hunk 15]
```diff
@@ -51,25 +57,60 @@ export const getDataFromDBStream = async (stream: NodeJS.ReadableStream, schema:
...
+};
+
+export const getTransactions = async (blockID: Buffer, db: KVStore) => {
+	try {
+		const txIDs: Buffer[] = [];
+		const ids = await db.get(`${DB_KEY_TRANSACTIONS_BLOCK_ID}:${keyString(blockID)}`);
+		const txIDLength = TRANSACTION_ID_LENGTH;
+		for (let i = 0; i < ids.length; i += txIDLength) {
+			txIDs.push(ids.slice(i, i + txIDLength));
+		}
```
Comment (priojeetpriyom) [inline src/createAsset.ts:96]: maybe we can get rid of the extra variable?

```suggestion
		for (let idx = 0; idx < ids.length; idx += TRANSACTION_ID_LENGTH) {
			txIDs.push(ids.slice(idx, idx + TRANSACTION_ID_LENGTH));
		}
```

**Review by sergeyshemyakov — APPROVED** [review]
Just one small comment from me, otherwise looks good!

**Review by priojeetpriyom — APPROVED** [review]

File: src/assets/pos.ts (hunk 14 above)
Comment (sameersubudhi) [reply src/assets/pos.ts:47]: The implementation is as per the LIP specifications and specific to the Lisk Core migration. Kindly do not modify!

File: src/assets/token.ts [hunk 16]
```diff
@@ -96,13 +97,16 @@ export const createUserSubstoreArray = async (
...
-				lockedBalances: await getLockedBalances(account),
-			};
-			userSubstore.push(userObj);
+			const lockedBalances = await getLockedBalances(account);
+			if (account.token.balance !== AMOUNT_ZERO || lockedBalances.length) {
+				const userObj = {
+					address: account.address,
+					tokenID: tokenIDBuffer,
+					availableBalance: String(account.token.balance),
+					lockedBalances,
```
[the remark this answers is not in the document]
Comment (sameersubudhi) [reply src/assets/token.ts:106]: In JS/TS an object with only a variable name takes the property name based on the variable name, i.e, both the examples below will result in the equivalent object:

```
{ lockedBalances } == { lockedBalances: lockedBalances }
```

**Pull request merged**